File: day4/src/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Card struct {
	CardNumber     int
	OwnNumbers     []int
	WinningNumbers []int
}

func parseInput(inputStr string) []Card {
	var cards []Card
	lines := strings.Split(strings.TrimSpace(inputStr), "\n")

	for _, line := range lines {
		parts := strings.Split(line, ": ")
		cardNumber, _ := strconv.Atoi(strings.Split(parts[0], " ")[1])

		numbers := strings.Split(parts[1], " | ")
		winningNumbers := convertToIntSlice(strings.Split(numbers[0], " "))
		ownNumbers := convertToIntSlice(strings.Split(numbers[1], " "))

		cards = append(cards, Card{cardNumber, ownNumbers, winningNumbers})
	}

	return cards
}

func getCardScore(card Card) int {
	score := 0
	for _, ownNumber := range card.OwnNumbers {
		for _, winningNumber := range card.WinningNumbers {
			if ownNumber == winningNumber {
				score++
			}
		}
	}
	return score
}

func solve(inputStr string) int {
	cards := parseInput(inputStr)
	counts := make([]int, len(cards))
	for i := range counts {
		counts[i] = 1
	}

	for i, card := range cards {
		score := getCardScore(card)
		if score > 0 {
			for j := i + 1; j < min(i+1+score, len(cards)); j++ {
				counts[j] += counts[i]
			}
		}
	}

	sum := 0
	for _, count := range counts {
		sum += count
	}
	return sum
}

func convertToIntSlice(strSlice []string) []int {
	var intSlice []int
	for _, str := range strSlice {
		num, _ := strconv.Atoi(str)
		intSlice = append(intSlice, num)
	}
	return intSlice
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	inputStr := ""
	for scanner.Scan() {
		inputStr += scanner.Text() + "\n"
	}

	start := time.Now()
	result := solve(inputStr)
	fmt.Printf("\nResult: %d\n", result)

	end := time.Now()
	fmt.Printf("Time taken: %d µs\n", end.Sub(start).Microseconds())
}

File: day4/src/main.py
from typing import List, NamedTuple
import time

class Card(NamedTuple):
    card_number: int
    own_numbers: List[int]
    winning_numbers: List[int]

def parse_input(input_str: str) -> List[Card]:
    cards = []
    for line in input_str.strip().split("\n"):
        parts = line.split(": ")
        card_number = int(parts[0].split()[1])

        numbers = parts[1].split(" | ")
        winning_numbers = [int(n) for n in numbers[0].split()]
        own_numbers = [int(n) for n in numbers[1].split()]

        cards.append(Card(card_number, own_numbers, winning_numbers))

    return cards

def get_card_score(card: Card) -> int:
    score = 0
    for own_number in card.own_numbers:
        for winning_number in card.winning_numbers:
            if own_number == winning_number:
                score += 1
    return score

def solve(input_str: str) -> int:
    cards = parse_input(input_str)
    counts = [1] * len(cards)

    for i, card in enumerate(cards):
        score = get_card_score(card)
        if score > 0:
            for j in range(i + 1, min(i + 1 + score, len(cards))):
                counts[j] += counts[i]

    return sum(counts)

def main():
    with open("input.txt") as f:
        input_str = f.read()

    start = time.time()
    result = solve(input_str)
    # println!("\nResult: {}", result);
    print(f"\nResult: {result}")

    end = time.time()
    # Print time in micros
    print(f"Time taken: {round((end - start) * 1000 * 1000)} µs")

if __name__ == "__main__":
    main()